[FEAT] -  Sage Icon Library implementation

ju-Skinner opened the pull request:
# Description
~~This is the initial commit for the Sage Icon library. The main focus right now is the `Figma-icon-export` script.~~ This PR is pretty big. I will suggest going through the commits to understand and see how things tie together. 

There is still a bit of documentation that needs to be completed. I hope this can be used as a good starting point. I can create additional tasks to address the missing needs.

* Improve Storybook documentation
* Incorporate the `cheatsheet.html` located in the dist folder into Storybook. This file list all of the available icons
* Confluence needs to be updated to show how to run the Figma Export process. 


## Updates
Added All Icosn to an available story and to the Docs page
![image](https://user-images.githubusercontent.com/1633290/214909464-8032d309-17f8-47cc-98eb-8bf9d823509c.png)

- Name control is now a select control, which includes all the available icons
![image](https://user-images.githubusercontent.com/1633290/214909631-7c1b88be-e610-4cb7-92f7-afe79e23db27.png)

Fixes #(issue)

## Type of change
- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

# How Has This Been Tested?

There are a few Unit and E2E at the moment. Most of the testing has been done using Storybook. I will need to improve the test coverage. 

- [x] unit tests
- [x] e2e tests
- [ ] accessibility tests
- [x] tested manually
- [x] other: Storybook

**Test Configuration**:

- sage versions:
- os:
- browsers:
- screen readers:
- misc:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

ju-Skinner committed:
feat(sage-icons): initial commit for sage-icons

ju-Skinner commented on libs/icons/readme.md:
````diff
@@ -0,0 +1,48 @@
+
+```sh
+yarn nx run icons:build
+```
````
As of right now, I believe this is fine. However, I believe this info should be for internal use.

ju-Skinner committed:
style(svg): add svg files that were pulled from figma

ju-Skinner committed:
chore(svg-optimize): script to optimize svgs

ju-Skinner committed:
chore: add eslint

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+# Sage Icons
+
+This is the Icon Library for Sage
```
```suggestion
This is the Icon Library for Sage.
```

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+
+### Creating a Figma access token
+1. Navigate to [Figma](https://www.figma.com) and login with your credentials.
+2. Click you Profile avatar and select 'Settings'
```
```suggestion
2. Click your Profile avatar and select 'Settings.'
```

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+1. Navigate to [Figma](https://www.figma.com) and login with your credentials.
+2. Click you Profile avatar and select 'Settings'
+3. Scroll down until you see 'Personal access tokens'
+4. Under 'Create a new personal access token:' enter a token description and hit enter
```
What would be a good example to write here?

ju-Skinner replied:
I used `Kajabi`, but it literally could have been anything.

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+2. Click you Profile avatar and select 'Settings'
+3. Scroll down until you see 'Personal access tokens'
+4. Under 'Create a new personal access token:' enter a token description and hit enter
+5. You should now see a token that you will need to copy and save somewhere safe
```
Is 1Password somewhere that we could suggest saving this token?

ju-Skinner replied:
Yes, we could save this Access Token there which would then be shared between the team and CI.

monicawheeler commented on libs/icons/readme.md:
````diff
@@ -0,0 +1,48 @@
+```
+{
+  "figmaAccessToken": "some-access-token",
+  "figmaFileId": "figma-file-id",
````
My `figmaFileId` is:
`#######/######/####3` 

 Is that correct for this field? Is it worth noting in the Attributes table something along the lines of, "The unique id will be a long string of mixed characters." Or something like that?

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+4. Under 'Create a new personal access token:' enter a token description and hit enter
+5. You should now see a token that you will need to copy and save somewhere safe
+
+### Configuration File
```
```suggestion
### Setting up the configuration file
```

Suggestion for the title, take it or leave it :)

monicawheeler commented on libs/icons/readme.md:
```diff
@@ -0,0 +1,48 @@
+
+**NOTE**: The `figmaAccessToken` and `figmaFileId` can be set in either the configuration file or as an environment variable. The environment variable will take precedence over the configuration file.
+
+### How to run
```
```suggestion
### Exporting icons
```

As before, suggesting a title update, take it or leave it :)

monicawheeler commented on libs/icons/scripts/readme.md:
```diff
@@ -0,0 +1,53 @@
+# Build and Development scripts
+
+# Updating Icons
+The `src/svg` directory is where the icons are downloaded to. These are originally pulled from the Figma document, which is considered to be the the single source of truth. These images are not optimized as they just exported out of the editor. In order to get the latest svgs, you will need run `yarn run nx icons:update`.
```
```suggestion
The `src/svg` directory is where the icons are downloaded. These are pulled initially from the Figma document, considered the single source of truth. These images are not optimized, as they are just exported from the editor. To get the latest svgs, you will need to run `yarn run nx icons:update`.
```

monicawheeler commented on libs/icons/scripts/readme.md:
```diff
@@ -0,0 +1,53 @@
+The `src/svg` directory is where the icons are downloaded to. These are originally pulled from the Figma document, which is considered to be the the single source of truth. These images are not optimized as they just exported out of the editor. In order to get the latest svgs, you will need run `yarn run nx icons:update`.
+
+# Figma Icon Configuration
+The `figma-icon-config.json` should be located in the root directory. If there isn't one there currently, you can follow the instructions below to create one. It is required in order to pull the icons from Figma.
```
```suggestion
The `figma-icon-config.json` should be located in the root directory. You can follow the instructions below to create one if there isn't one currently. The presence of this file is required to pull icons from Figma.
```

monicawheeler commented on libs/icons/scripts/readme.md:
```diff
@@ -0,0 +1,53 @@
+
+### Creating a Figma access token
+1. Navigate to [Figma](https://www.figma.com) and login with your credentials.
+2. Click you Profile avatar and select 'Settings'
```
```suggestion
2. Click your Profile avatar and select 'Settings'
```

monicawheeler commented on libs/icons/scripts/readme.md:
```diff
@@ -0,0 +1,53 @@
+**NOTE**: The `figmaAccessToken` and `figmaFileId` can be set in either the configuration file or as an environment variable. The environment variable will take precedence over the configuration file.
+
+# Building locally
+Once the latest svgs have been update, which would either be added, deleted, or modified in the `src/svg` directory, you can run:
```
```suggestion
Once the latest svgs have been updated, which would either be added, deleted, or modified in the `src/svg` directory, you can run:
```

monicawheeler commented on libs/icons/scripts/readme.md:
````diff
@@ -0,0 +1,53 @@
+  yarn run nx icons:build
+```
+
+The `build` command will both download the latest icons from the figma document (see [Figma Icon Configuration](#figma-doc-configuration)), optimized all of the icons, and generate the files to be distributed. Once complete and successful all of the optimized svgs will be located in `dist/sage-icons/svg`, along with the `sage-icon` web component located in the `dist` directory.
````
```suggestion
The `build` command will download the latest icons from the Figma document (see [Figma Icon Configuration](#figma-doc-configuration)), optimize all of the icons, and generate the files to be distributed. Once complete and successful, all of the optimized svgs will be located in `dist/sage-icons/svg`, along with the `sage-icon` web component in the `dist` directory.
```

ju-Skinner committed:
feat(sage-icon): implementation for the sage-icon

ju-Skinner committed:
feat(sage-icon): updates to copy svgs to dist/collection/components/sage-icon folder

ju-Skinner committed:
chore(icons): add test.spec.watch command

ju-Skinner committed:
feat(core): add sage-icon library to core lib

chore(storybook): merge icon and core docs.json files

This is needed to have a single source to list out

the properties for the components using argsTable function

ju-Skinner committed:
feat(core): include globalScript for Core library

ju-Skinner committed:
feat(sage-react): add the 'sage-icon' library to the project

ju-Skinner committed:
chore(ci): add build actions to build core and icon in a composite build lifecycle

ju-Skinner committed:
chore(setup): update start command to include icon library

ju-Skinner committed:
chore(setup): change dependencies versions and targets

this builds the correct dependency graph for the current libraries

teenwolfblitzer commented on libs/icons/src/components/sage-icon/sage-icon.tsx:
```diff
@@ -0,0 +1,151 @@
+        {...inheritedAttributes}
+      >
+        {Build.isBrowser && this.svgContent ? (
+          <div innerHTML={this.svgContent}></div>
```
Do we have the ability to use empty JSX fragments (in place of the `div`) in these kinds of cases?

teenwolfblitzer commented on libs/icons/figma-icon-config.json:
```diff
@@ -0,0 +1,6 @@
+{
```
Thoughts on adding this file to the `gitignore`? Could help to prevent accidental commits of personal access tokens… which I may or may not have done before 😓

ju-Skinner replied:
That's a good thought and question. We don't want to ignore this file because it is dependent. Access tokens should be in the .env file.

teenwolfblitzer reviewed (approved):
👍 LGTM, working great! Thanks for doing this, it'll be a huge improvement on our current (manual) process.

pixelflips reviewed (approved):
LGTM! Working great and all parties involved in icon updates are going to be thrilled! Great work!

anechol reviewed (approved):
So happy to have this working. This is gonna save us quite a lot of time and effort. Is that something that can be measured?

goodwinchris commented on libs/icons/src/components/sage-icon/sage-icon.tsx:
```diff
@@ -0,0 +1,151 @@
+   * The name of the icon to use from
+   * the built-in set.
+   */
+  @Prop({ reflect: true }) name?: string;
```
Would be great to get a populated list of icon names here

ju-Skinner replied:
That's more of a feature implementation for Storybook not the component.

goodwinchris commented on libs/icons/scripts/cheatsheet-template.html:
```diff
@@ -0,0 +1,41 @@
+<!doctype html>
```
testing info on how to view this page would be great

ju-Skinner replied:
It renders when you run the project with `yarn nx run icons:start`

goodwinchris reviewed (approved):
Things look great with my spot testing! I left a few comments about "nice to haves" and PR testing but those are non-blockers and can happen in follow-ups. 

🚀

ju-Skinner committed:
feat(sage-icon): update storybook documentation

includes populated list of icons of Default story

Adds story All Icons which list all the available icons

ju-Skinner committed:
feat(sage-icon): remove color attribute for later development

ju-Skinner committed:
chore(docs): update sage-icon story to use IconGallery component

ju-Skinner committed:
chore(ci): add build dependency for devs when using start.stencil command

ju-Skinner committed:
refactor(sage-icon): change the 'normal' size to use 'regular'
